### PR TITLE
Quick sort algorithms 

### PR DESCRIPTION
**Pull request by BriseBalloches**
### **Describe your change:**

Improve the two quick sort algorithms (simple and 3 partitions).
Added type hints, docstrings, doctests, better implementation when possible.
Since the two files concern the same algorithm (one being a variation of the other), i decided to create one PR.

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [X] Documentation change?

### **Checklist:**
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: sorts/quick_sort_3_partition.py [hunk 1]
```diff
@@ -1,25 +1,63 @@
...
+"""
+This is a pure Python implementation of the three partition quick sort algorithm
+
+This algorithm splits at each iteration a mutable ordered collection in 3 parts
+1 : elements smaller than the pivot
+2 : elements equal to the pivot (not happening in 'simple' quick sort)
+3 : elements greater than the pivot
+
+For doctests run following command:
+python -m doctest -v quick_sort_3partition.py
```
Comment (cclauss) [inline sorts/quick_sort_3_partition.py:10]:
```suggestion
```

File: sorts/quick_sort_3_partition.py [hunk 2]
```diff
@@ -1,25 +1,63 @@
...
+or
```
Comment (cclauss) [inline sorts/quick_sort_3_partition.py:11]:
```suggestion
```

File: sorts/quick_sort_3_partition.py [hunk 3]
```diff
@@ -1,25 +1,63 @@
...
+python3 -m doctest -v quick_sort_3partition.py
+
+For manual testing run:
+python quick_sort_3partition.py
+"""
+
+
+def quick_sort_3partition(my_list: list, left: int = 0, right: int = None) -> list:
```
Comment (cclauss) [inline sorts/quick_sort_3_partition.py:19]: You cannot declare a variable to be an int and then assign None to it.

File: sorts/quick_sort_3_partition.py [hunk 4]
```diff
@@ -1,25 +1,63 @@
...
-        return
-    a = i = left
-    b = right
-    pivot = sorting[left]
-    while i <= b:
-        if sorting[i] < pivot:
-            sorting[a], sorting[i] = sorting[i], sorting[a]
-            a += 1
+        return my_list
+    i: int = left
```
Comment (cclauss) [inline sorts/quick_sort_3_partition.py:37]: left is already declared to be an int so type checkers are smart enough to know that I is an int.

**Comment by BriseBalloches** [comment]
Updated with your comments. Thanks.

**Comment by nelphax** [comment]
Unscribe

**Pull request closed without merging**